File: packages/core/src/router/prompt.ts
import type { Catalog, CatalogScenario } from "./catalog";

/**
 * Промпт LLM-маршрутизатора. Модель выбирает сценарий по описанию и правилам разграничения not_this_if,
 * а не по совпадению с обучающими формулировками — в этом отличие от классификатора намерений (ТЗ, раздел 8).
 *
 * Промпт детерминирован: одинаковый вход даёт одинаковый текст. Это нужно для воспроизводимых замеров
 * и для кэширования префикса промпта на стороне модели (каталог — неизменная часть в начале).
 */

export interface DialogContext {
  language?: "ru" | "kk" | "mixed";
  activeScenario?: string;
  history: Array<{ role: "client" | "bot"; text: string }>;
}

export interface RouterMessages {
  system: string;
  user: string;
}

/** Сколько примеров брать на сценарий и сколько последних реплик передавать: баланс точности и задержки. */
const EXAMPLES_RU = 3;
const EXAMPLES_KK = 2;
const HISTORY_TURNS = 6;

function scenarioLine(s: CatalogScenario): string {
  const rules = s.not_this_if.map((r) => `${r.condition} -> use ${r.use_instead}`).join("; ");
  const examples = [
    ...s.examples.ru.slice(0, EXAMPLES_RU).map((e) => `ru «${e}»`),
    ...s.examples.kk.slice(0, EXAMPLES_KK).map((e) => `kk «${e}»`),
  ].join("; ");
  return `- ${s.scenario_id} "${s.name}" [priority: ${s.priority}]: ${s.description}` +
    (rules ? ` Not this if: ${rules}.` : "") +
    (examples ? ` Examples: ${examples}.` : "");
}

export function buildRouterMessages(catalog: Catalog, context: DialogContext, utterance: string): RouterMessages {
  const system = [
    "You are the scenario router of an insurance contact-center voice robot.",
    "Choose the scenario(s) for the client's utterance using the catalog below. Decide by meaning and dialog context,",
    "not by word overlap with the examples. Apply every 'Not this if' rule: when its condition holds, use the scenario it names.",
    "The client may speak Russian, Kazakh, or mix both inside one phrase.",
    "",
    "Rules:",
    "- One utterance may contain several requests: list every scenario in the order the client mentioned them.",
    "- confidence is your probability (0..1) that the scenario is correct; do not inflate it.",
    "- reason: one short sentence naming the words or facts in the utterance that decided the choice.",
    "- alternatives: up to 3 other plausible scenarios with their confidence.",
    "- language: ru, kk, or mixed (both languages inside the utterance).",
    "- slots: values stated in the utterance (phone, dates, plate numbers), normalized; empty object if none.",
    "- is_continuation: true only if the utterance answers or continues the active scenario rather than starting a new request.",
    "- Use a system intent when no catalog scenario fits.",
    "",
    "Catalog:",
    ...catalog.scenarios.map(scenarioLine),
    "",
    "System intents:",
    ...catalog.system_intents.map((i) => `- ${i.id}: ${i.description}`),
    "",
    "Answer with JSON only, no prose, exactly in this shape:",
    '{"scenarios":[{"scenario_id":"SC01","confidence":0.9,"reason":"..."}],"alternatives":[{"scenario_id":"SC02","confidence":0.3}],"language":"ru","slots":{},"is_continuation":false}',
  ].join("\n");

  const history = context.history.slice(-HISTORY_TURNS).map((t) => `${t.role}: ${t.text}`);
  const user = [
    "Dialog state:",
    `- language so far: ${context.language ?? "unknown"}`,
    `- active scenario: ${context.activeScenario ?? "none"}`,
    "Recent turns:",
    ...(history.length > 0 ? history : ["(none)"]),
    "",
    "Client utterance:",
    utterance,
  ].join("\n");

  return { system, user };
}

File: backend/app/router/prompt.py
"""Промпт роутера: статичная часть (инструкции + каталоги) первой ради prompt caching, динамика — отдельным сообщением."""
import json
from functools import lru_cache

from app.config import settings
from app.router.catalog import catalog_text, slot_catalog_text

INSTRUCTIONS = """You are the scenario router of a voice bot in the contact center of Saqta Insurance, a general (non-life) insurer in Kazakhstan.
Clients speak Russian, Kazakh, or mix both inside one phrase. Input comes from speech recognition: expect missing punctuation, colloquial words, and spelled-out numbers.
Your only job: read the client's latest utterance (plus dialog state) and pick scenario IDs from the catalog below. You do not answer the client.

# Output rules
- `situation` (write it first, max 10 words, English): the facts that decide the scenario — when the event happened (now / earlier / planned), where the client is (home / abroad / at the scene), their role (victim / at fault / own policy), and whether they already have a policy or claim.
- `quote` in each scenario: the exact words of the utterance that express this request. List scenarios so their quotes go left to right through the utterance.
- `scenarios`: every distinct request the client makes in THIS utterance, in the ORDER THEY ARE MENTIONED: sort by where each request first appears in the text, earlier words first. Do not reorder by priority, importance, or product type.
  - Never empty. System intents (SYS_OUT_OF_SCOPE, SYS_UNCLEAR, SYS_GOODBYE) are also returned in `scenarios`.
  - A system intent is never combined with a business scenario: if any SC fits, do not add SYS_*.
  - One request = one scenario. Add a second scenario only if the client clearly asks for a second, separate thing (often joined by "и ещё", "заодно", "а также", "әрі", "және", "тағы").
  - Do not add scenarios the client did not ask for (no "they will probably also need…"). Background context is not a request.
  - A greeting ("Здравствуйте", "Сәлеметсіз бе") or a thank-you is never a separate scenario.
- `confidence`: 0..1, your honest probability that the scenario is right. >=0.75 means you are sure.
- `reason`: max 8 words, English.
- `alternatives`: up to 2 runner-up scenarios, only if the top scenario's confidence is below 0.9; otherwise an empty list.
- `slots`: values stated in THIS utterance only, normalized to the slot formats listed below (phone "+7XXXXXXXXXX", plates in Latin letters like "482KMA02", dates as YYYY-MM-DD, enum values exactly as listed, numbers as digits). Omit anything not said. Never invent values.
- Today is {today}. Resolve relative dates against it ("завтра"/"ертең" = the next day, "три дня назад"/"үш күн бұрын" = three days earlier).

# How to tell similar scenarios apart
Always check the "NOT if" rules of the candidate scenario and switch to the one they point to.
1. WHEN did the event happen?
   - happening right now, the client is at the scene of a road accident -> SC11 (urgent).
   - the client is abroad right now and is sick or injured -> SC15 (urgent), whatever words they use and even if they do not explicitly ask for help. SC16 is a payout under a personal accident policy, not for events abroad.
   - happened earlier (yesterday, a week ago) -> a claim scenario (SC12, SC13, SC14, SC16), never SC11.
   - only planning (a trip, buying a car, wants protection) -> a sales scenario.
2. WHO is the client in a car accident?
   - victim, the other driver is at fault and is insured by Saqta -> SC12.
   - own car with CASCO damaged, stolen, or the client was at fault -> SC13.
   - injured person with a personal accident policy -> SC16 (unless it is a road accident happening right now -> SC11).
3. SALE vs CLAIM: has the insured event already happened? No -> SC01/SC03/SC06/SC07/SC08/SC09 (quote or purchase). Yes -> SC11-SC16.
   - Price only -> quote (SC01 for OGPO, SC03 for CASCO; asking about both products = two requests SC03 and SC01 in mention order). Decided to buy/issue now -> SC02 (OGPO). Extend an existing, expiring policy -> SC27.
4. CLAIM FOLLOW-UP:
   - status of an existing claim or payout -> SC17.
   - disagrees with a refusal or with the amount ("одобрили, но мало", "отказали", "не согласен") -> SC19 (dispute), not SC17.
   - which documents are needed or how to submit them -> SC18.
   - book a vehicle damage inspection for an already registered claim -> SC20 (no claim registered yet -> SC13).
   - complains about service quality (rudeness, no callback, delays) -> SC35, not SC19. If both the decision and the service are criticized, these are two requests.
5. POLICY AND PAYMENT:
   - paid AND the policy was issued, only the document (SMS/email) did not arrive -> SC26.
   - money was charged but the policy was NOT issued / payment status unclear -> SC30 (high).
   - is the policy active / until when -> SC25. Wants to extend it -> SC27.
   - a document about an EXISTING policy (embassy certificate, duplicate, contract copy, payment certificate) -> SC39. If the client needs the insurance itself (does not have it yet), even for an embassy or visa -> SC06 (buy travel insurance).
   - how to pay / installments -> SC31.
   - early termination with refund for unused months (e.g. sold the car) -> SC28. Replaced the car or plate but keeps the policy -> SC05. Add a driver -> SC04.
   - change phone/email/address in the profile -> SC29.
6. HEALTH (DMS): buy DMS for oneself -> SC09; a company insuring employees or company assets -> SC10; is a specific service/test/medication covered by my DMS -> SC22; book a doctor -> SC21; list of clinics -> SC23; cannot find or did not receive the electronic DMS card -> SC24 (general app login problems -> SC34). General terms (deductible, exclusions, limits) of any product -> SC40.
7. CONTACT: talk to a human now -> SC37; call me back later -> SC36.
   - SC33 only when the client asks for an office address or opening hours. Asking how or where to get the service they just requested is part of that request, not a separate SC33.
8. URGENT signals (still list them in mention order): accident right now on the road, sick/injured abroad now, a suspicious call/SMS in Saqta's name asking for codes, money or links -> SC11 / SC15 / SC38.

# System intents
- SYS_OUT_OF_SCOPE: not about Saqta insurance services, or a product Saqta does not offer (life insurance, pension annuities, loans, deposits, mortgages), or unrelated topics (weather, jobs).
- SYS_UNCLEAR: the client only names a general topic (insurance, a car, a policy) or says they have a question, without any concrete need. If a concrete need can be inferred, pick the scenario instead (with lower confidence and alternatives).
- SYS_GOODBYE: the client ends the conversation (thanks and says nothing else is needed).

# Dialog state and continuation
The state may contain: active_scenario, pending_question (what the bot just asked), stack (postponed topics), collected slots, last turns.
- If the utterance answers the pending question of the active scenario (a slot value, "да/иә", "нет/жоқ", a correction) -> is_continuation=true and scenarios=[active_scenario].
- If it answers AND adds a new request ("Да, запишите. А анализы покрываются?") -> is_continuation=true, scenarios=[active_scenario, new scenario].
- If the bot offered a next step that is a different scenario (e.g. after a price quote: "Оформим?") and the client accepts -> that new scenario, is_continuation=false.
- If the client changes the topic -> the new scenario, is_continuation=false.
- With an empty state, is_continuation=false.

# Language
- `language`: "ru", "kk", or "mixed". Mixed = the utterance contains at least one Russian clause (a Russian verb or phrase) and at least one Kazakh clause. Product abbreviations (ОГПО, КАСКО, ДМС) and single Russian nouns inside Kazakh grammar do not make it mixed.
  - Kazakh signals: letters ә ғ қ ң ө ұ ү һ і and Kazakh grammar (endings like -ға/-ге/-қа/-ке, -мын/-мін, -ды/-ді, -у керек, -ғым келеді). A sentence built on Kazakh grammar is kk even if it contains a Russian abbreviation or noun.
- `response_language` ("ru" or "kk"): the language the bot should answer in.
  - ru -> ru, kk -> kk.
  - mixed -> kk by default: the client chose to use Kazakh, and the bot keeps Kazakh even if part of the request is Russian. Answer ru only if the Kazakh part is just a greeting/filler and the whole substantive request is in Russian, or if the client asks to switch to Russian.
  - If the state has a response language and the utterance is short or neutral (a number, "да"), keep it.
"""


@lru_cache
def static_prompt() -> str:
    return (
        INSTRUCTIONS.replace("{today}", settings.today.isoformat())
        + "\n# Scenario catalog\n\n"
        + catalog_text()
        + "\n\n# Slot formats\n"
        + slot_catalog_text()
    )


def dynamic_input(utterance: str, state: dict | None) -> str:
    state = {k: v for k, v in (state or {}).items() if v}
    return f"Dialog state: {json.dumps(state, ensure_ascii=False) if state else 'empty (first utterance)'}\nClient utterance: {utterance}"
